library/Icingadb/Data/PivotTable.php

```diff
@@ -0,0 +1,410 @@
+     *
+     * @return  $this
+     */
+    public function setXAxisFilter(Filter\Rule $filter = null)
+    {
+        $this->xAxisFilter = $filter;
+        return  $this;
+    }
+
+    /**
+     * Set the filter to apply on the query for the y-axis
+     *
+     * @param   Filter  $filter
+     *
+     * @return  $this
+     */
+    public function setYAxisFilter(Filter\Rule $filter = null)
+    {
+        $this->yAxisFilter = $filter;
+        return  $this;
+    }
+
+    /**
+     * Get the x-axis header
+     *
+     * Defaults to {@link $xAxisColumn} in case no x-axis header has been set using {@link setXAxisHeader()}
+     *
+     * @return string
+     */
+    public function getXAxisHeader()
+    {
+        return $this->xAxisHeader !== null ? $this->xAxisHeader : $this->xAxisColumn;
+    }
+
+    /**
+     * Set the x-axis header
+     *
+     * @param   string $xAxisHeader
+     *
+     * @return  $this
+     */
+    public function setXAxisHeader($xAxisHeader)
+    {
+        $this->xAxisHeader = (string) $xAxisHeader;
+        return $this;
+    }
+
+    /**
+     * Get the y-axis header
+     *
+     * Defaults to {@link $yAxisColumn} in case no x-axis header has been set using {@link setYAxisHeader()}
+     *
+     * @return string
+     */
+    public function getYAxisHeader()
+    {
+        return $this->yAxisHeader !== null ? $this->yAxisHeader : $this->yAxisColumn;
+    }
+
+    /**
+     * Set the y-axis header
+     *
+     * @param   string $yAxisHeader
+     *
+     * @return  $this
+     */
+    public function setYAxisHeader($yAxisHeader)
+    {
+        $this->yAxisHeader = (string) $yAxisHeader;
+        return $this;
+    }
+
+    /**
+     * Return the value for the given request parameter
+     *
+     * @param   string  $axis       The axis for which to return the parameter ('x' or 'y')
+     * @param   string  $param      The parameter name to return
+     * @param   int     $default    The default value to return
+     *
+     * @return  int
+     */
+    protected function getPaginationParameter($axis, $param, $default = null)
+    {
+        $request = Icinga::app()->getRequest();
+
+        $value = $request->getParam($param, '');
+        if (strpos($value, ',') > 0) {
+            $parts = explode(',', $value, 2);
+            return intval($parts[$axis === 'x' ? 0 : 1]);
+        }
+
+        return $default !== null ? $default : 0;
+    }
+
+    /**
+     * Query horizontal (x) axis
+     *
+     * @return Query
+     */
+    protected function queryXAxis()
+    {
+        if ($this->xAxisQuery === null) {
+            $this->xAxisQuery = clone $this->baseQuery;
+            $xAxisHeader = $this->getXAxisHeader();
+            $table = $this->xAxisQuery->getModel()->getTableName();
+            $columns = array_unique([
+                $this->xAxisColumn => $this->gridcols[$this->xAxisColumn],
+                $xAxisHeader => $this->gridcols[$xAxisHeader]
+            ]);
```

Comment:
`array_unique` shouldn't be applied to the select columns. I've suggested to use PDO's fetchPairs functionality below. This won't work with less than two select columns. The group columns however don't need to contain duplicates, there `array_unique` is appropriate.